_posts/2023-10-04-#28107.md

```diff
@@ -7,17 +7,55 @@ authors: [dergoegge]
-
-<!-- TODO: Before meeting, add notes and questions
 ## Notes
 
+- Since the activation of SegWit, transactions are identifiable by two types of identifiers: `txid` and `wtxid`. Both identifiers are a `uint256` resulting from a double `sha256` hash of serialized transaction data. The core [difference](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/primitives/transaction.h#L276-L280) is that `wtxid` includes the witness data, while `txid` does not.
+ 
+- Type-safety is a concept in programming to help avoid type errors, enforcing strict correspondence between a variable and its type. It ensures that a program performs only legal operations on data, with validation occurring during compile-time. This proactive error detection and prevention lead to more reliable and maintainable software, minimizing unintended behavior and vulnerabilities. An example of an earlier type-safety improvement in Bitcoin Core is the incorporation of `std::chrono`, introduced in [#16908](https://github.com/bitcoin/bitcoin/pull/16908).
+
+- In the context, of transaction identifiers, type-safety is important because it is impossible to deduce the type difference from just the `uint256` value. With `GenTxid` type checking can be done, but it is more verbose and done at run-time instead of at compile-time. Additionally, transaction identifier types improve readability of code by being explicit about expected types instead of solely relying on clear variable naming.
+
+- This PR aims to improve type safety by introducing the new [`transaction_identifier`](https://github.com/bitcoin-core-review-club/bitcoin/commit/28f9ff3d5416407d3c6d5374f6fc3c2767f4a7f3#diff-f3d91988bf7d98529d638fa829dc3fab695be002305a7614eaaeb283d4fb8101R18) type, with `Txid` and `Wtxid` typedefs for brevity.
+
+- There are instances where `wtxid` and `txid` are treated identically, such as for example in [`most_recent_block_txs`](https://github.com/bitcoin/bitcoin/blob/6619d6a8dca5a4d8e664a76526ac6bef3eb17831/src/net_processing.cpp#L1916-L1918), where the `uint256` type is used to handle this. In other contexts such as [`CTxMemPool::info`](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/txmempool.cpp#L862) where `wtxid` and `txid` are treated differently but either is accepted, the [`GenTxid`](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/primitives/transaction.h#L425) type is adopted.
+
 ## Questions
+
+### Concept
+
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+1. What does it mean for a transaction identifier to be type-safe? Why is that important or helpful? Are there any downsides?
+
+1. Conceptually, when writing new code that requires referring to transactions, when should you use `txid` and when should you use `wtxid`?
```

Comment:
Question idea: Can you point to any examples in the code where using one instead of the other could be very bad?

Reply:
1 point per example. Good students will come with lots of examples

Reply:
Yes, added!